google/cloud/spanner_dbapi/parse_utils.py

```diff
@@ -0,0 +1,546 @@
+        if (len(params) % n_stride) != 0:
+            raise ProgrammingError(
+                "Invalid length: len(params)=%d MUST be a multiple of "
+                "len(pyformat_args)=%d" % (len(params), n_stride)
+            )
+
+    # Now chop up the strides.
+    strides = []
+    for step in range(0, len(params), n_stride):
+        stride = tuple(params[step : step + n_stride :])
+        strides.append(stride)
+
+    return strides
+
+
+def sql_pyformat_args_to_spanner(sql, params):
+    """
+    Transform pyformat set SQL to named arguments for Cloud Spanner.
+    It will also unescape previously escaped format specifiers
+    like %%s to %s.
+    For example:
+        SQL:      'SELECT * from t where f1=%s, f2=%s, f3=%s'
+        Params:   ('a', 23, '888***')
+    becomes:
+        SQL:      'SELECT * from t where f1=@a0, f2=@a1, f3=@a2'
+        Params:   {'a0': 'a', 'a1': 23, 'a2': '888***'}
+
+    OR
+        SQL:      'SELECT * from t where f1=%(f1)s, f2=%(f2)s, f3=%(f3)s'
+        Params:   {'f1': 'a', 'f2': 23, 'f3': '888***', 'extra': 'aye')
+    becomes:
+        SQL:      'SELECT * from t where f1=@a0, f2=@a1, f3=@a2'
+        Params:   {'a0': 'a', 'a1': 23, 'a2': '888***'}
+    """
+    if not params:
+        return sanitize_literals_for_upload(sql), params
+
+    found_pyformat_placeholders = RE_PYFORMAT.findall(sql)
+    params_is_dict = isinstance(params, dict)
+
+    if params_is_dict:
+        if not found_pyformat_placeholders:
+            return sanitize_literals_for_upload(sql), params
+    else:
+        n_params = len(params) if params else 0
+        n_matches = len(found_pyformat_placeholders)
+        if n_matches != n_params:
+            raise Error(
+                "pyformat_args mismatch\ngot %d args from %s\n"
+                "want %d args in %s"
+                % (n_matches, found_pyformat_placeholders, n_params, params)
+            )
+
+    named_args = {}
+    # We've now got for example:
+    # Case a) Params is a non-dict
+    #   SQL:      'SELECT * from t where f1=%s, f2=%s, f3=%s'
+    #   Params:   ('a', 23, '888***')
+    # Case b) Params is a dict and the matches are %(value)s'
+    for i, pyfmt in enumerate(found_pyformat_placeholders):
+        key = "a%d" % i
+        sql = sql.replace(pyfmt, "@" + key, 1)
+        if params_is_dict:
+            # The '%(key)s' case, so interpolate it.
+            resolved_value = pyfmt % params
+            named_args[key] = resolved_value
+        else:
+            named_args[key] = cast_for_spanner(params[i])
+
+    return sanitize_literals_for_upload(sql), named_args
+
+
+def cast_for_spanner(value):
+    """Convert the param to its Cloud Spanner equivalent type.
+
+    :type value: Any
+    :param value: Value to convert to a Cloud Spanner type.
+
+    :rtype: Any
+    :returns: Value converted to a Cloud Spanner type.
+    """
+    if isinstance(value, decimal.Decimal):
+        return float(value)
```

Comment:
Converting `Decimal` to `float` can lose precision. Please use `str` instead. 

If you have a good reason for not doing so, you will instead need to clearly document that NUMERICs may lose precision when using this driver.

Reply:
Changed as suggested, based on this [recommendation](https://cloud.google.com/spanner/docs/storing-numeric-data#recommendation_store_arbitrary_precision_numbers_as_strings).